text/0000-fabric-private-chaincode.md

```diff
@@ -0,0 +1,472 @@
+
+## Deployment Process (in detail)
+
+This section details the turn-up process for all elements of FPC, including an explanation of the trust architecture.
+
+* Step 1: Channel and Peers
+
+	Initially it is assumed that a standard Fabric network is in place. A new Channel is created according to the existing standard process for Fabric. One or more Peers join the channel according to the standard process.
+	Using the `peer channel join` command, an FPC-enabled endorsing peer attaches a Ledger Enclave to the Channel, imprinting itself with a content-addressable Channel Identifier and a hash of the Channel's Genesis Block contents.
+	Note that FPC currently does not support cross-channel interaction; however, this can be addressed in future work.
+
+* Step 2: Create the FPC Chaincode deployment artifact
+
+	One or more of the participants compiles an agreed-upon chaincode using the FPC compiler configuration.
+	This links in the FPC Shim, including its shim in the compiled package and creating a unique chaincode identifier called the MRENCLAVE.
+	Details of this process will vary in future versions supporting other TEEs. Participants must be careful to use the same compiler configuration and version numbers of all included software. As long as they do, the MRENCLAVE will be the same for each compiled Chaincode Package.
+
+* Step 3: Install and Approve an FPC Chaincode
+
+	Administrators in each organization that will participate in the channel install the agreed-upon chaincode package on their FPC-enabled Peers following the standard Fabric 2.0 lifecycle process.
+	They perform the installation of the chaincode tarball onto their Peers and invoke an `approveformyorg` transaction specifying MRENCLAVE (recall Step 1) as the chaincode version.
+	Thereby, the organization approves a particular FPC Chaincode by defining the corresponding chaincode identity (MRENCLAVE).
+	Once enough Approvals have been committed in order to fulfill the Lifecycle Endorsement Policy, an admin commits the Chaincode Definition by invoking `peer lifecycle chaincode commit`.
+	Note that by defining the MRENCLAVE inside the Chaincode Definition, FPC enforces that only enclaves matching the defined MRENCLAVE can produce valid transactions of the given chaincode.
+
+* Step 4: Enclave Creation and Attestation Generation
+
+	Once the FPC chaincode installation and approval is complete, the admin invokes `peer lifecycle chaincode createenclave` at an endorsing peer.
+    Thereby, the FPC Shim creates an new Chaincode Enclave running the FPC Chaincode.
+    In particular, when creating the enclave, the FPC Shim also fetches the MRENCLAVE and Channel ID of the Ledger Enclave and passes them with the create command.
+	Now, having all necessary information, the new Enclave proceeds with the key generation protocol.
+	It creates public/private key-pairs for signing and encryption. The public signing key is also used as enclave identifier.
```

Comment:
> It creates public/private key-pairs for signing and encryption.

How does this work in the case of several enclave instances? 
Like, you have an enclave in your bank, but i want to have an enclave in my bank, so i need your encryption key as well. 
Since key generation relies on secret entropy, one way or another one of us needs to transfer the other some random bytes. How do we do that? Do we have some kind of infrastructure to do key exchange? Do we have a global system setup of bootstrap encryption keys?

Reply:
See the `fpc-registration` and `fpc-key-dist` UML diagrams referenced further down in the text which gives the details of the involved protocols.

Reply:
Is there a paper that describes this UML flow in English? If I zoom out the text is unreadable, and if i zoom in i see like 1% of the entire diagram so I can't figure out what's going on...

Reply:
There is no paper, but the source files (PUML) are fairly readable text files. 
On a high-level as mentioned above the first enclave creates all chaincode secrets and then will distribute it to other enclaves encrypted using the other enclave's encryption key. The key of course is to make sure that only authorized enclaves, i.e., enclaves running the correct code, and with a consistent view on channel identity, chaincode definition and trusted ledger identity get the key.  The devil for security, though, is really in the detail which right means the UML diagrams ...